StreamChatClient_v3/Controllers/ChannelListController.swift

```diff
@@ -5,91 +5,251 @@
+public protocol ChannelListControllerDelegate: ControllerRemoteActivityDelegate {
+    func controller(_ controller: ChannelListControllerGeneric<DefaultDataTypes>, didChangeChannels changes: [Change<Channel>])
+}
+
+public extension ChannelListControllerDelegate {
+    func controller(_ controller: ChannelListControllerGeneric<DefaultDataTypes>, didChangeChannels changes: [Change<Channel>]) {}
+}
+
+/// `ChannelListController` uses this protocol to communicate changes to its delegate.
+///
+/// If you're **not** using custom extra data types, you can use a convenience version of this protocol
+/// named `ChannelListControllerDelegate`, which hides the generic types, and make the usage easier.
+public protocol ChannelListControllerDelegateGeneric: ControllerRemoteActivityDelegate {
+    associatedtype ExtraData: ExtraDataTypes
+    func controller(
+        _ controller: ChannelListControllerGeneric<ExtraData>,
+        didChangeChannels changes: [Change<ChannelModel<ExtraData>>]
+    )
+}
+
+public extension ChannelListControllerDelegateGeneric {
+    func controller(_ controller: ChannelListControllerGeneric<DefaultDataTypes>,
+                    didChangeChannels changes: [Change<ChannelModel<ExtraData>>]) {}
+}
+
+extension ClientError {
+    public class FetchFailed: Error {
+        public var localizedDescription: String = "Failed to perform fetch request. This is an internal error."
+    }
+}
+
+// MARK: - Delegate type eraser
+
+class AnyChannelListControllerDelegate<ExtraData: ExtraDataTypes>: ChannelListControllerDelegateGeneric {
+    private var _controllerDidChangeChannels: (ChannelListControllerGeneric<ExtraData>, [Change<ChannelModel<ExtraData>>]) -> Void
+    private var _controllerWillStartFetchingRemoteData: (Controller) -> Void
+    private var _controllerDidStopFetchingRemoteData: (Controller, Error?) -> Void
     
-    private let worker: ChannelQueryUpdater<ExtraData>
-    private let viewContext: NSManagedObjectContext
+    weak var wrappedDelegate: AnyObject?
     
-    private lazy var fetchResultsController: NSFetchedResultsController<ChannelDTO> = {
-        let request = ChannelDTO.channelListFetchRequest(query: self.query)
-        return .init(fetchRequest: request,
-                     managedObjectContext: viewContext,
-                     sectionNameKeyPath: nil,
-                     cacheName: nil)
-    }()
+    init(
+        wrappedDelegate: AnyObject?,
+        controllerWillStartFetchingRemoteData: @escaping (Controller) -> Void,
+        controllerDidStopFetchingRemoteData: @escaping (Controller, Error?) -> Void,
+        controllerDidChangeChannels: @escaping (ChannelListControllerGeneric<ExtraData>, [Change<ChannelModel<ExtraData>>]) -> Void
+    ) {
+        _controllerWillStartFetchingRemoteData = controllerWillStartFetchingRemoteData
+        _controllerDidStopFetchingRemoteData = controllerDidStopFetchingRemoteData
+        _controllerDidChangeChannels = controllerDidChangeChannels
```

Comment:
`wrappedDelegate` is not assigned

Reply:
Thanks!! I actually saw a failing test because of this but somehow convinced myself it the test was wrong 🤦‍♂️

Reply:
Fixed in 29a5dacc78fd07a577c91b876e40cf20f5954fd6

Reply:
Since we have wrappedDelegate here, why do we need to pass the delegate functions separately? we can just call `delegate?.function`

Reply:
I wanted to have both flows unified. `ChannelListController` doesn't know about the convenience delegate with default extra data. If you removed the extension, everything would still work.

Reply:
It's really just convenience API if it's implemented like this